<?xml version="1.0"?>

<!--
Launches Scenario 1 - Navigation 0: , No navigation, just blind movement.
This should Crash!
-->

<launch>
  <!-- PARAMS ...................................................... -->
  <arg name="gazeboGui" default="true"/>
  <arg name="rvizGui" default="true"/>
  <arg name="robot_id" default="1"/>

  <arg name="scenario_id" default="1"/>
  <!-- scenario_id == 1 Robot crossess left of human         -->
  <!-- scenario_id == 2 Robot crossess right of human        -->
  <!-- scenario_id == 3 Robot crossess is overtaken by human -->
  <!-- scenario_id == 4 Robot crossess crosses human         -->

  <arg name="navigation_id" default="0"/>
  <!-- navigation_id == 0 No navigation stack  -->
  <!-- navigation_id == 1 DWA Navigation Stack -->
  <!-- navigation_id == 2 TEB Navigation Stack -->
  <!-- navigation_id == 3 MPC Navigation Stack -->



  <!-- DERIVED ARGS ................................................ -->
  <!-- TF prefix of robot -->
  <arg name="prefix" default="robot$(arg robot_id)"/>

  <arg name="descriptor" value="S$(arg scenario_id)-N$(arg navigation_id)"/>

  <arg name="max_scenario_time" value="30" if="$(eval arg('navigation_id') == 0)"/>
  <arg name="max_scenario_time" value="60" if="$(eval arg('navigation_id') == 1)"/>
  <arg name="max_scenario_time" value="60" if="$(eval arg('navigation_id') == 2)"/>
  <arg name="max_scenario_time" value="60" if="$(eval arg('navigation_id') == 3)"/>

  <arg name="useTwist" value="true"  if="$(eval arg('navigation_id') != 3)"/>
  <arg name="useTwist" value="false" if="$(eval arg('navigation_id') == 3)"/>

  <arg name="rviz_config_file" value="$(find nav_simulation)/rviz/iliad_cititruck_dwa_nav.rviz" if="$(eval arg('navigation_id') == 0)"/>
  <arg name="rviz_config_file" value="$(find nav_simulation)/rviz/iliad_cititruck_dwa_nav.rviz" if="$(eval arg('navigation_id') == 1)"/>
  <arg name="rviz_config_file" value="$(find nav_simulation)/rviz/iliad_cititruck_teb_nav.rviz" if="$(eval arg('navigation_id') == 2)"/>
  <arg name="rviz_config_file" value="$(find nav_simulation)/rviz/iliad_cititruck_mpc_nav.rviz" if="$(eval arg('navigation_id') == 3)"/>



  <arg name="navigation_launch_file" value="$(find nav_simulation)/launch/move_base_dwa.launch" if="$(eval arg('navigation_id') == 1)"/>
  <arg name="navigation_launch_file" value="$(find nav_simulation)/launch/move_base_teb.launch" if="$(eval arg('navigation_id') == 2)"/>
  <arg name="navigation_launch_file" value="$(find nav_simulation)/launch/move_base_mpc.launch" if="$(eval arg('navigation_id') == 3)"/>

  <!-- SET UP GAZEBO WORLD AND VISUALIZATION ....................... -->
  <include file="$(find base_simulation)/launch/scenario_setup.launch">
      <arg name="scenario_id" value="$(arg scenario_id)"/>
      <arg name="robot_id" value="$(arg robot_id)"/>
      <arg name="prefix" value="$(arg prefix)"/>
      <arg name="gazeboGui" value="$(arg gazeboGui)"/>
      <arg name="rvizGui" value="$(arg rvizGui)"/>
      <arg name="rviz_config_file" value="$(arg rviz_config_file)"/>
      <arg name="cmd_vel_topic" value="control/controller/cmd_vel" />
      <arg name="useTwist" value="$(arg useTwist)"/>
      <arg name="map_frame" value="/map_laser2d"/>
      <arg name="map_topic" value="/maps/map_laser2d"/>
  </include>

  <!-- SET UP NAVIGATION ............................................-->
  <include file="$(arg navigation_launch_file)" if="$(eval arg('navigation_id') > 0)">
      <arg name="robot_id" value="$(arg robot_id)"/>
      <arg name="prefix" value="$(arg prefix)"/>
      <!-- "relative" base_frame -->
      <arg name="base_frame" value="base_link"/>
      <arg name="map_frame" value="/map_laser2d"/>
      <arg name="map_topic" value="/maps/map_laser2d"/>
  </include>

  <!-- Things to record: collisions, robot position, goal achieved?, human position  -->
  <node pkg="robot_pose_publisher" type="robot_pose_publisher" name="steer_pose_publisher" output="screen">
    <param name="map_frame" value="world"/>
    <param name="base_frame" value="$(arg prefix)/base_footprint"/>
    <param name="is_stamped" value="true"/>
    <remap from="robot_pose" to="/$(arg prefix)/robot_pose"/>
  </node>

  <node pkg="taros19_experiments" type="gazebo_model_pose_publisher.py" name="actor_pose_publisher" output="screen">
    <param name="map_frame" value="world"/>
    <param name="model_name" value="actor1"/>
  </node>

  <node required="true" pkg="rosbag" type="record" name="recorder"
    args="record --duration=$(arg max_scenario_time) -o $(arg descriptor) /isColliding /$(arg prefix)/robot_pose /$(arg prefix)/control/report  /$(arg prefix)/move_base/status /actor_pose " output="screen"/>

</launch>
